Snowplow/Internal/Tracker/SPTrackerControllerImpl.m

```diff
@@ -285,6 +285,9 @@ - (NSString *)version {
 // MARK: - Private methods
 
 - (SPTracker *)tracker {
+    if (![self.serviceProvider isTrackerInitialized]) {
+        SPLogError(@"Recreating tracker instance after it was removed. This will not be supported in future versions.");
+    }
```

Comment:
Checking if the tracker is initialised is good but a bit implementation specific. I wonder if we can be even more explicit. When they remove the tracker the TrackerController instance should disappear but unfortunately this can't happen as they have the reference. If we add an internal lifecycle state in TrackerController we may explicitly mark it as "zombie" (or so). We may invalidate tracking operations on a zombie controller and raise errors when they are used. Actually, the ServiceProvider should be marked as zombie or even better we could explicitly remove the instance of a ServiceProvider when a TrackerController is removed. So even the ServiceProvider can be destroyed. Doing that, the TrackerController would return nil for almost all the method. IMO it's not a breaking change because it's a fix of a bug. A dead TrackerController shouldn't be used so it wouldn't be part of the API we support.
I understand this can be arguable. I think it's ok with this implementation for a quick fix. In a future release, I would consider to explicitly invalid controllers (at least the TrackerController) when it's removed. Unfortunately, it's something that should have been implemented already at v2.

Reply:
I agree with this and it was also my original intention to return null from the TrackerController if it is a zombie one (or raise an error), but the problem is that all methods in the TrackerController currently return nonnull result (e.g., getting the namespace). We would need to change the signature of the methods to let them return null and it could have a significant impact on the user code since at least in Swift and Kotlin, they need to explicitly handle optional types.

I think for now we can only do what this PR is doing, but we should make the properties nullable or let them raise errors in a future major version (or the rewrite).